src/userservice/tests/test_userservice.py

```diff
@@ -0,0 +1,205 @@
+        self.flask_app.config['VERSION'] = version
+        # send get request to test client
+        response = self.test_app.get('/version')
+        # assert both versions are equal
+        self.assertEqual(response.data, version.encode())
+
+    def test_ready_endpoint(self):
+        """test if correct response is returned from readiness probe"""
+        response = self.test_app.get('/ready')
+        self.assertEqual(response.data, b'ok')
+
+    def test_create_user(self):
+        """test creating a new user who does not exist in the DB"""
+        # mock return value of get_user which checks if user exists as None
+        self.mocked_db.return_value.get_user.return_value = None
+        # create example user request
+        example_user_request = EXAMPLE_USER_REQUEST.copy()
+        example_user_request['username'] = 'foo'
+        # send request to test client
+        response = self.test_app.post('/users', data=example_user_request)
+        # assert 201 response code
+        self.assertEqual(response.status_code, 201)
+
+    def test_create_user_existing(self):
+        """test creating a new user who already exists in the DB"""
+        # mock return value of get_user which checks if user exists
+        self.mocked_db.return_value.get_user.return_value = {}
+        example_user_request = EXAMPLE_USER_REQUEST.copy()
+        # create example user request
+        example_user_request['username'] = 'foo'
+        # send request to test client
+        response = self.test_app.post('/users', data=example_user_request)
+        # assert 409 response code
+        self.assertEqual(response.status_code, 409)
+        # assert we get correct error message
+        self.assertEqual(
+            response.json['msg'], 'user {} already exists'.format(example_user_request['username'])
+        )
+
+    def test_create_user_sql_error(self):
+        """test creating a new user but throws SQL error when trying to add"""
+        # mock return value of get_user which checks if user exists as None
+        self.mocked_db.return_value.get_user.return_value = None
+        # mock return value of add_user to throw SQLAlchemyError
+        self.mocked_db.return_value.add_user.side_effect = SQLAlchemyError()
+        # create example user request
+        example_user = EXAMPLE_USER_REQUEST.copy()
+        example_user['username'] = 'foo'
+        # send request to test client
+        response = self.test_app.post('/users', data=example_user)
+        # assert 500 response code
+        self.assertEqual(response.status_code, 500)
+        # assert we get correct error message
+        self.assertEqual(response.json['msg'], 'failed to create user')
+
+    def test_create_user_malformed(self):
+        """test creating a new user without required keys"""
+        # create example user request
+        example_user = EXAMPLE_USER_REQUEST.copy()
+        # remove a required field
+        example_user.pop('username')
```

Comment:
Added this - not sure if you meant use 1 object and keep removing keys from it or to try removing a different key each time. I implemented the latter since I thought it was a better test but can switch to former as well.

Reply:
Yeah I was also thinking of the later, looks good!